Documentation/contributing/development/debugging.rst

```diff
@@ -0,0 +1,308 @@
+            rules:
+              dns:
+                - matchPattern: "*"
+      - toFQDNs:
+          - matchPattern: "*"
+
+The L7 DNS rule has an implicit L3 allow-all because it defines only L4 and L7
+sections. This is the second selector in the list, and includes all possible L3
+identities known in the system. In contrast, the first selector, which
+corresponds to the ``toFQDNS: matchName: "*"`` rule would list all identities
+for IPs that came from the DNS Proxy. Other CIDR identities would not be
+included.
+
+*Datapath Plumbing*
+For a policy to be fully realized the datapath for an Endpoint must be updated.
+In the case of a new DNS-source IP, the CIDR identity associated with it must
+propagate from the selectors to the Endpoint specific policy. Unless a new
+policy is being added, this often only involves updating the Policy Map of the
+Endpoint with the new CIDR Identity of the IP. This can be verified:
+
+.. code:: bash
+
+    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get 3459
+    DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS
+    Ingress     reserved:unknown              ANY          NONE         1367    7
+    Ingress     reserved:host                 ANY          NONE         0       0
+    Egress      reserved:unknown              53/TCP       36447        0       0
+    Egress      reserved:unknown              53/UDP       36447        138     2
+    Egress      cidr:104.198.14.52/32         ANY          NONE         477     6
+                reserved:world 
+
+Note that the labels for identities are resolved here. This can be skipped, or there may be cases where this doesn't occur:
+
+.. code:: bash
+
+    $ kubectl exec pod/cilium-sbp8v -n cilium -- cilium bpf policy get -n 3459
+    DIRECTION   IDENTITY   PORT/PROTO   PROXY PORT   BYTES   PACKETS
+    Ingress     0          ANY          NONE         1367    7
+    Ingress     1          ANY          NONE         0       0
+    Egress      0          53/TCP       36447        0       0
+    Egress      0          53/UDP       36447        138     2
+    Egress      16777217   ANY          NONE         477     6
+
+
+L3 ``toFQDNs`` rules are egress only, so we would expect to see an ``Egress``
+entry with Security Identity ``16777217``. The L7 rule, used to redirect to the
+DNS Proxy is also present with a populated ``PROXY PORT``. It has a 0
+``IDENTITY`` as it is an L3 wildcard.
+
+An identity missing here can be an error in various places:
+- Policy doesn't actually allow this Endpoint to connect. A sanity check is to
+use ``cilium endpoint list`` to see if cilium thinks it should have policy
+enforcement.
+- Endpoint regeneration is slow and the Policy Map has not been updated yet.
+This can occur in cases where we have leaked IPs from the DNS cache (i.e. the
+were never deleted correctly) or when there are legitimately many IPs. It can
+also simply mean an overloaded node or even a deadlock within cilium.
```

Comment:
In this case, is the deadlock in cilium you are suggesting the result of resource constraints?